woltka/workflow.py

```diff
@@ -0,0 +1,590 @@
+    elif isdir(path_):
+
+        # turn off demultiplexing if not decided
+        demux = demux or False
+
+        # get a map of plausible sample Ids to files
+        map_ = id2file_map(path_, ext, not demux and samples)
+        if len(map_) == 0:
+            raise ValueError('No valid file found in directory.')
+        if demux:
+            files = sorted([join(path_, x) for x in map_.values()])
+
+        # validate with given sample Ids
+        else:
+            if not samples:
+                samples = sorted(map_.keys())
+            elif len(map_) < len(samples):
+                raise ValueError(errmsg)
+            files = {join(path_, map_[x]): x for x in samples}
+
+    else:
+        raise ValueError(f'"{path_}" is not a valid file or directory.')
+
+    click.echo(f'Alignment files to read: {len(files)}.')
+    click.echo(f'Demultiplexing: {"on" if demux else "off"}.')
+    return samples, files, demux
+
+
+def build_align_proc(coords_fp: str = None,
+                     overlap:   int = None) -> object:
+    """Build alignment processor.
+
+    Parameters
+    ----------
+    coords_fp : str, optional
+        Path to gene coordinates file.
+    overlap : int, optional
+        Read/gene overlapping percentage threshold
+
+    Returns
+    -------
+    object
+        Alignment processor.
+
+    Notes
+    -----
+    Currently two processors are supported: Plain() for regular alignments
+    (i.e., simple query-to-subject maps), Ordinal() for alignments with
+    coordinates which will be used to match queries (reads) and genes. The
+    presence of a gene coordinates file (`coords_fp`) is an indicator for
+    using the latter.
+    """
+    if coords_fp:
+        click.echo(f'Reading gene coordinates...', nl=False)
+        with readzip(coords_fp) as fh:
+            coords = read_gene_coords(fh, sort=True)
+        click.echo(' Done.')
+        click.echo(f'Total host sequences: {len(coords)}.')
+        return Ordinal(coords, whether_prefix(coords),
+                       overlap and overlap / 100)
+    else:
+        return Plain()
+
+
+def build_hierarchy(names_fp:   str = None,
+                    nodes_fp:   str = None,
+                    newick_fp:  str = None,
+                    lineage_fp: str = None,
+                    ranktb_fp:  str = None,
+                    map_fps:   list = None,
+                    map_rank:  bool = False) -> (dict, dict, dict, str):
+    """Construct hierarchical classification system.
+
+    Parameters
+    ----------
+    names_fp : str, optional
+        Taxonomic names file.
+    nodes_fp : str, optional
+        Taxonomic nodes file.
+    newick_fp : str, optional
+        Newick tree file.
+    lineage_fp : str, optional
+        Lineage strings file.
+    ranktb_fp : str, optional
+        Rank table file.
+    map_fps : list of str, optional
+        Mapping files.
+    map_rank : bool, optional
+        Mapping filename is rank.
+
+    Returns
```

Comment:
See the numpy docs on formatting `Returns` in docstring:
https://docs.scipy.org/doc/numpy/docs/howto_document.html#docstring-standard
Namely, multiple outputs should be specified as follows:
```
Returns
-------
err_code : int
    Non-zero value indicates error code, or zero on success.
err_msg : str or None
    Human readable error message, or None on success.
```
I tend to agree with this format as more readable.

Reply:
For this part I am not clear. I thought that the format you quoted only applies when the return value is an object with attributes which can be directly called. For example when you do `res = pearsonr(arr1, arr2)`, you can then do `res.statistic` and `res.pvalue` in addition to `res[0]` and `res[1]`. However in the current scenario where the function returns a simple tuple, this format may not apply. What do you think?

Reply:
Sorry I may rephrase (more explictly) to perhaps do something like
```
dict
    Taxonomic tree.
dict
    Rank dictionary
dict
    Name dictionary
str
    Root identifier
```
I think this convention makes it very explicit what each type corresponds to. And it makes things less confusing if you were to do something like add longer descriptions. e.g.,

```
dict
    Taxonomic tree. And now I am going to give it a really really
    really really really really really really really really really really 
    really really really really really really long description.
dict
    Rank dictionary
dict
    Name dictionary
str
    Root identifier
```

is more readable than

```
dict, dict, dict, str
    Taxonomic tree. And now I am going to give it a really really
        really really really really really really really really really really 
        really really really really really really long description.
    Rank dictionary
    Name dictionary
    Root identifier
```

Reply:
Good point! Will change this as well as others.